chia/full_node/full_node_api.py

```diff
@@ -1720,5 +1712,201 @@ def get_fee_estimates(est: FeeEstimatorInterface, req_times: List[uint64]) -> Li
+            return msg
+
+        min_height = uint32((request.previous_height + 1) if request.previous_height is not None else 0)
+
+        # This is a limit imposed by `batch_coin_states_by_puzzle_hashes`, due to the SQLite variable limit.
+        # It can be increased in the future, and this protocol should be written and tested in a way that
+        # this increase would not break the API.
+        count = 15000
+        puzzle_hashes = request.puzzle_hashes[:count]
+
+        (coin_states, next_min_height) = await self.full_node.coin_store.batch_coin_states_by_puzzle_hashes(
+            puzzle_hashes,
+            min_height=min_height,
+            include_spent=request.filters.include_spent,
+            include_unspent=request.filters.include_unspent,
+            include_hinted=request.filters.include_hinted,
+            min_amount=request.filters.min_amount,
+            max_items=max_items,
+        )
+        is_done = next_min_height is None
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        height = uint32(next_min_height - 1) if next_min_height is not None else peak_height
+        header_hash = self.full_node.blockchain.height_to_hash(height)
+        assert header_hash is not None
+
+        if is_done and request.subscribe_when_finished:
+            added = list(
+                self.full_node.subscriptions.add_puzzle_subscriptions(
+                    peer.peer_node_id, puzzle_hashes, max_subscriptions
+                )
+            )
+        else:
+            added = puzzle_hashes
+
+        response = wallet_protocol.RespondPuzzleState(added, height, header_hash, is_done, coin_states)
+        msg = make_msg(ProtocolMessageTypes.respond_puzzle_state, response)
+        return msg
+
+    @api_request(peer_required=True, reply_types=[ProtocolMessageTypes.respond_coin_state])
+    async def request_coin_state(self, request: wallet_protocol.RequestCoinState, peer: WSChiaConnection) -> Message:
+        max_items = self.max_subscribe_response_items(peer)
+        max_subscriptions = self.max_subscriptions(peer)
+
+        previous_header_hash = (
+            self.full_node.blockchain.height_to_hash(request.previous_height)
+            if request.previous_height is not None
+            else self.full_node.blockchain.constants.GENESIS_CHALLENGE
+        )
+        assert previous_header_hash is not None
+
+        if request.header_hash != previous_header_hash:
+            rejection = wallet_protocol.RejectCoinState()
+            msg = make_msg(ProtocolMessageTypes.reject_coin_state, rejection)
+            return msg
+
+        min_height = uint32(request.previous_height + 1 if request.previous_height is not None else 0)
+
+        coin_states = await self.full_node.coin_store.get_coin_states_by_ids(
+            True,
+            set(request.coin_ids),
+            min_height=min_height,
+            max_items=max_items,
```

Comment:
if the response is truncated here, we don't do anything to tell the wallet or anything. This looks like a potential source of corrupt wallet balances.

Reply:
The wallet can determine which coin ids were omitted on its own, technically. Perhaps it would make sense though to return only the coin ids that were included?

Reply:
the logic here looks a bit fishy. It seems like we end up subscribing, even when we're not "done" sending all the state. It doesn't look like there's any message indicating this (like a subscription response message).

Part of my confusion here is that coin IDs are fundamentally different than puzzle hashes.

Reply:
What do you mean by done? There isn't any batching involved with `RequestCoinState` since you're always going to get a number of coin states at most equal to the number of coin ids you ask for in the response.

Reply:
right. it's the requestor's responsibility to not ask for more coin states than what fit in the response. If you request a coin ID that doesn't exist yet, you still subscribe to it, so you're notified if it appears. Right?

When subscribe means that if there's a reorg, and the coin disappears or re-appears or is spent, you get notified.

It still seems a bit difficult for the requestor to understand what happened if a request hits against `max_items`. You won't get the coin state, even though you requested it. Do you interpret that as if the coin not existing or as the message was truncated?

Reply:
This has been fixed - if the request is higher than max items it will fail truncate it, and the returned coin ids will represent everything which was actually queried (deduplicated). If it exceeds the subscription limit the request will fail now.